### DISCUSSION TITLE
Why "21030018" convert to float normal  , but "21030019" shown as scientific notation

### DISCUSSION THREAD

**Discussion by BruceQiu1996** [dotnet/csharplang#7746] [Q&A]
i can't get answer from anywhere.

platfrom : .net 7
![微信图片_20231207160545](https://github.com/dotnet/csharplang/assets/28331315/075f91e9-ee6b-4be2-a42a-facf06d6502a)
![微信图片_20231207160549](https://github.com/dotnet/csharplang/assets/28331315/c2777ed1-e69d-41c2-b13e-e708fe537ac1)

**Comment by yaakov-h** [comment]
This seems like an IDE issue and not a language issue, however you haven't even mentioned what IDE you're using.

**Reply by BruceQiu1996** [reply]
visual studio 2022

**Comment by TahirAhmadov** [comment]
`float` is a floating point number, which means it's stored in an in-precise manner from the get-go. In your case, the way the number is converted to the underlying bits changes when you go from 21030018 to 21030019 - for the former, it's able to preserve all the digits, and for latter, it's not. The precision is specified to be 6-9 digits, and your number is straddling that line, which explains this behavior.
https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/floating-point-numeric-types
If you are worried about preserving all the digits, you should use `decimal` (or even `int`/`long`), not `float`/`double`.
PS. This probably belongs more to Stack Overflow and not C# language design board.

**Comment by huoyaoyuan** [comment]
20130019 is exceeding the precision of float.
```csharp
        float a1 = 21030018f;
        Console.WriteLine(a1);

        float a2 = 21030019f;
        Console.WriteLine(a2);
```
outputs on .NET Core:
```
21030018
21030020
```

However, I have no idea why the two values show differently in VS debugger. `ToString` uses both normal form on .NET Core, and both scientific notation on .NET Framework.

**Reply by BruceQiu1996** [reply]
Hello, can you tell me why the 21030019 exceeded the accuracy, but 21030018 did not. I can't understand it

**Reply by jnm2** [reply]
https://www.h-schmidt.net/FloatConverter/IEEE754.html may help you visualize what the 32 bits are doing to hold the two numbers.
